### ISSUE TITLE
suggest Single Abstract Method

### ISSUE THREAD

**Issue by zmunm** [detekt/detekt#3528]
## Expected Behavior of the rule

SAM conversion is possible, but I would like to find out where I am creating an anonymous class.

java interface
```java
public interface Foo {
    void bar();
}
```

Written
```kotlin
val listener = object : Foo {
    override fun bar() {
         ...
    }
}
```

What I want

```kotlin
val listener = Foo {
    ...
}
```

## Context

https://kotlinlang.org/docs/java-interop.html#sam-conversions

**Comment by chao2zhang** [comment]
Hello @zmunm, this is a good idea since IntelliJ/Android Studio has the inspection to convert to the SAM syntax "Redundant SAM constructor - This inspection reports SAM constructor usages which can be replaced with lambdas". If you would like to contribute, please let us know.

**Comment by zmunm** [comment]
It may take a while, but I want to try :)

**Comment by cortinico** [comment]
I'm assigning this to you @zmunm 👍 Feel free to let us know if you need further support.

**Comment by zmunm** [comment]
I have met some edge cases and have had a hard time. 😞 

TL; DR
Because of the default implementation of the interface, it is difficult to check if it is a SAM by looking at only the file.
Very close example is [here](https://github.com/detekt/detekt/blob/master/detekt-core/src/test/kotlin/io/gitlab/arturbosch/detekt/core/TopLevelAutoCorrectSpec.kt)

---
Long description

There is a syntax that looks like SAM. But not always.
```kotlin
object: Sam {
    override fun foo() {
    }
}
```


non compliant java interface
```java
interface Sam {
    default void foo() {}
}
```


non compliant kotlin interface
```kotlin
interface Sam {
    fun foo() {}
}
```


compliant java interface
```java
interface Sam {
    void foo();
    default void bar() {}
}
```


compliant kotlin interface
```kotlin
fun interface Sam {
    fun foo()
    fun bar() {}
}
```

Is there a good way to check this?

**Comment by cortinico** [comment]
> Is there a good way to check this?

I'm unsure as I never faced such scenario.
Ideally you want to access the `Sam` interface during the inspection and check if it's a SAM interface or not 🤔

**Comment by zmunm** [comment]
@cortinico 
It seems that I can try to check the package and import by iterating through the `Class.forName`.
However, I doubt that using reflect is acceptable for detekt.

+++
There were a few more cases, and I think it's hard to see without reflect.

1. overloading
```kotlin
interface A {
    fun foo(a: Boolean = true)
}

object : B {
    override fun foo(a: Boolean) {
    }
}
```

2. implement chain
```kotlin
interface A {
    fun foo()
}

fun interface B : A

B {}
```

**Comment by cortinico** [comment]
> It seems that I can try to check the package and import by iterating through the `Class.forName`.
> However, I doubt that using reflect is acceptable for detekt.

Yup that's definitely a blocker.

By any chance is the Kotlin Compiler or IntelliJ providing any working for this?

**Comment by zmunm** [comment]
> By any chance is the Kotlin Compiler or IntelliJ providing any working for this?

It gave me a hint and thanks to that, I found a way. Thank you.
